sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -160,6 +161,42 @@ case class Expm1(child: Expression) extends UnaryMathExpression(math.expm1, "EXP
 
 case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR")
 
+case class Factorial(child: Expression)
+  extends UnaryExpression with Serializable with ExpectsInputTypes {
+  override def expectedChildTypes: Seq[DataType] = Seq(IntegerType)
+  override def dataType: DataType = IntegerType
+  override def foldable: Boolean = child.foldable
+  override def nullable: Boolean = true
+  override def toString: String = s"factorial($child)"
+  override def eval(input: InternalRow): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null) {
+      null
+    } else {
+      val input = evalE.asInstanceOf[Integer]
+      if (input > 20 || input < 0) {
+        sys.error(s"Valid input for factorial is [0..20], but the current input is: $input")
+      }
+      IntMath.factorial(input)
```

Comment:
gonna to cache a result array in our factorial method and get rid of the guava one.